### PR TITLE
Feat/validate refer reward

### PR DESCRIPTION
**Pull request by lmscunha**
Updates:

- Inside the actAcceptEntry action, add a new row in the reward table. 
As it stands, for every one entry that is accepted, the reward points increases by 1.

- Add code format instructions to Readme.

- This PR lean on the approval of [PR nº5](https://github.com/senecajs/seneca-refer/pull/5)

- Close issue: #7

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "url": "git://github.com/senecajs/seneca-refer.git"
   },
   "scripts": {
+    "format": "prettier --write .",
```
Comment (rjrodger) [inline package.json:19]: rename this to `prettier` as this is the convention used in other seneca plugins

File: src/refer.ts [hunk 2]
```diff
@@ -81,32 +100,44 @@ function refer(this: any, options: any) {
...
-
 }
 
+async function getEntryReward(seneca: any, entry: any): Promise<object> {
+  const reward = await seneca.entity('refer/reward').load$({
+    entry_id: entry.id
+  })
+
+  if (!reward) {
+    return await seneca.entity('refer/reward').save$({
```
Comment (rjrodger) [inline src/refer.ts:115]: let's make this a little more DRY: create the entity, but do not save, and set count to 0. Do not return, drop through to lines 123,124 to do the save

File: src/refer.ts [hunk 3]
```diff
@@ -81,32 +100,44 @@ function refer(this: any, options: any) {
...
+      entry_kind: entry.kind,
+      entry_id: entry.id,
+      kind: 'accept',
+      count: 1
+    })
+  }
+
+  reward.count = reward.count + 1
+  return await seneca.entity('refer/reward').save$(reward)
```
Comment (rjrodger) [inline src/refer.ts:124]: `reward.save$()` is better
see https://senecajs.org/docs/tutorials/understanding-data-entities.html

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (LeonNimoy) [reply package.json]: Understood. Thanks for highlight this. Changes made on this commit: https://github.com/senecajs/seneca-refer/pull/6/commits/02a6a70a2483d1ece7d90527422f22fb5256d6ab#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R19

File: src/refer.ts (hunk 2 above)
Comment (LeonNimoy) [reply src/refer.ts]: Thank you for the suggestion. I updated this script, please take a look again.

File: src/refer.ts (hunk 3 above)
Comment (LeonNimoy) [reply src/refer.ts]: Indeed! I followed your suggestion, please take a look again.

**Comment by rjrodger** [comment]
The basic.messages test just tests one referral and reward. We need a `multi.messages` test (create the test file and copy the setup for basic.messages) that tests multiple users making multiple referrals. We will also need to handle the case were a referral is sent to the same person (perhaps use a `conflict.messages` for that to keep it separate)

File: README.md [hunk 4]
```diff
@@ -75,4 +84,84 @@ No description provided.
 
 ## Contributing
 
+## Code Formatter
```
Comment (rjrodger) [inline README.md:87]: put this in a separate doc file: CONTRIBUTING.md - most users will not be editing the code

File: src/refer.ts [hunk 5]
```diff
@@ -7,33 +7,47 @@ function refer(this: any, options: any) {
...
-    const occur = await seneca.entity('refer/occur').save$({
-      user_id: msg.user_id,
-      entry_kind: msg.kind,
+    let occur = await seneca.entity('refer/occur').load$({
       email: msg.email,
-      entry_id: entry.id,
-      kind: 'create',
+      kind: 'accept',
     })
+    if (!occur) {
```
Comment (rjrodger) [inline src/refer.ts:21]: use whitespace to create conceptual blocks - see https://wiredcraft.com/blog/robin-williams-four-basic-design-principles-for-non-designers/ *Proximity*
Thus: `if` should probably be separated by a line from the load$

File: src/refer.ts [hunk 6]
```diff
@@ -7,33 +7,47 @@ function refer(this: any, options: any) {
...
+      }
+    }
     return {
-      ok: true,
-      entry,
-      occur: [occur],
+      ok: false,
+      why: 'entry-invalid',
+      details: {
+        why_exactly: 'entry already a user',
```
Comment (rjrodger) [inline src/refer.ts:49]: don't do this - `details` is for specific info  - say `email`

File: src/refer.ts (hunk 5 above)
Comment (rjrodger) [inline src/refer.ts:21]: better to prefer `true` conditionals in general
in this case, also better to exit early if the data is bad - this avoids an extra level of indentation:

```
if (failing_condition) {
  return { ok: false, ... }
}
```

File: src/refer.ts [hunk 7]
```diff
@@ -7,33 +7,47 @@ function refer(this: any, options: any) {
...
+        entry,
+        occur: [occur],
+      }
+    }
     return {
-      ok: true,
-      entry,
-      occur: [occur],
+      ok: false,
+      why: 'entry-invalid',
```
Comment (rjrodger) [inline src/refer.ts:47]: perhaps `entry-exists`? that seems to be the actual issue?

File: src/refer.ts [hunk 8]
```diff
@@ -64,6 +90,52 @@ function refer(this: any, options: any) {
...
   }
 
+  async function actLostEntry(this: any, msg: any) {
+    const seneca = this
+
+    const entryList = await seneca.entity('refer/occur').list$({
+      email: msg.email,
+      kind: 'create',
+    })
+    entryList.forEach((entry: any) => {
```
Comment (rjrodger) [inline src/refer.ts:100]: in general, place seneca operations inside for loops, so you can use await inside them to ensure all operations complete before returning

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (LeonNimoy) [reply README.md]: Indeed, done!

File: src/refer.ts (hunk 5 above)
Comment (LeonNimoy) [reply src/refer.ts:22]: Thanks for let me know. Done!

File: src/refer.ts (hunk 6 above)
Comment (LeonNimoy) [reply src/refer.ts]: Got it. Done!

File: src/refer.ts (hunk 5 above)
Comment (LeonNimoy) [reply src/refer.ts]: Indeed. Done!

File: src/refer.ts (hunk 7 above)
Comment (LeonNimoy) [reply src/refer.ts]: Sounds good to me. Done!

File: src/refer.ts (hunk 8 above)
Comment (LeonNimoy) [reply src/refer.ts]: Indeed, thanks for let me know. Done!

File: src/refer.ts [hunk 9]
```diff
@@ -64,6 +90,55 @@ function refer(this: any, options: any) {
...
 
+  async function actLostEntry(this: any, msg: any) {
+    const seneca = this
+
+    const entryList = await seneca.entity('refer/occur').list$({
+      email: msg.email,
+      kind: 'create',
+    })
+
+    for (let i = 0; i < entryList.length; i++) {
```
Comment (rjrodger) [inline src/refer.ts:101]: to support longer term maintainability, let's refactor this. Use Array.filter on entryList to select only the entries that will be updated, then loop over just those entries, removing the conditional from the for loop (`if` statements bad :) ):

```
const chooseAGoodNameForThisFilteredList =
  entryList.filter(entry=>entryList[i].user_id !== msg.userWinner)
```

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: src/refer.ts (hunk 9 above)
Comment (LeonNimoy) [reply src/refer.ts]: Indeed, thanks for the suggestion.

For this variable name, I choose unacceptedReferrals. What do you think? Changes on commit: 7c1ffb3ae0925129087cd0e22c6e24405f00838d

**Pull request merged**